sympy/calculus/codomain.py

```diff
@@ -0,0 +1,166 @@
+
+    if not func.is_rational_function(symbol):
+        raise NotImplementedError("Algorithms finding range for non-rational functions"
+                                    "are not yet implemented")
+
+    if domain.is_EmptySet:
+        return EmptySet()
+
+    if not func.has(symbol):
+        return FiniteSet(func)
+
+    # this block of code can be replaced by
+    # sing = Intersection(FiniteSet(*sing), domain.closure)
+    # after the issue #9706 has been fixed
+    def closure_handle(set_im, singul):
+        if set_im.has(Interval):
+            if not oo in set_im.boundary:
+                if not S.NegativeInfinity in set_im.boundary:
+                    return Intersection(FiniteSet(*singul), set_im.closure)
+                return Intersection(FiniteSet(*singul), Union(set_im,
+                                    FiniteSet(max(set_im.boundary))))
+            else:
+                if not S.NegativeInfinity in set_im.boundary:
+                    return Intersection(FiniteSet(*singul), Union(set_im,
+                                        FiniteSet(min(set_im.boundary))))
+                return Intersection(FiniteSet(*singul), set_im)
+        return Intersection(FiniteSet(*singul), set_im)
+
+    # all the singularities of the function
+    sing = singularities(func, symbol)
+    sing_in_domain = closure_handle(domain, sing)
+
+    def codomain_interval(f, set_val, *sym):
+        symb = sym[0]
+        df1 = diff(f, symb)
+        df2 = diff(df1, symb)
+        der_zero = solveset_real(df1, symb)
+        der_zero_in_dom = closure_handle(set_val, der_zero)
+
+        maxi = set()
+        mini = set()
+        start_val = limit(f, symb, set_val.start)
+        end_val = limit(f, symb, set_val.end, '-')
+
+        for i in sing_in_domain:
+            if not i in set_val.boundary:
+                return Union(codomain(f, Interval(set_val.start, i, set_val.left_open, True), symb),
+                            codomain(f, Interval(i, set_val.end, True, set_val.right_open), symb))
+
+        if start_val is S.Infinity or end_val is S.Infinity:
+            maxi = set([(oo, True)])
+        elif start_val is S.NegativeInfinity or end_val is S.NegativeInfinity:
+            mini = set([(-oo, True)])
+        if maxi == set():
+            if start_val > end_val:
+                maxi = set([(start_val, set_val.left_open)])
+            elif start_val < end_val:
+                maxi = set([(end_val, set_val.right_open)])
+            else:
+                maxi = set([(start_val, set_val.left_open and set_val.right_open)])
+        if mini == set():
+            if start_val < end_val:
+                mini = set([(start_val, set_val.left_open)])
+            elif start_val > end_val:
+                mini = set([(end_val, set_val.right_open)])
+            else:
+                mini = set([(start_val, set_val.left_open and set_val.right_open)])
+
+        unk = set()
```

Comment:
OK. I will change it to `d2f_zero_val` though earlier it meant `unknown for d2f`.